### PR TITLE
feat(ng-dev): support using no targeting labels for merge

### PR DESCRIPTION
**Pull request by josephperrott**
Allow the configuration of no target labels instead always merging into the main branch provided in the `GithubConfig`.

### CODE REVIEW

File: ng-dev/pr/merge/pull-request.ts [hunk 1]
```diff
@@ -85,15 +85,17 @@ export async function loadAndValidatePullRequest(
...
+    const commitsInPr = prData.commits.nodes.map((n) => parseCommitMessage(n.commit.message));
 
-  try {
-    assertPendingState(prData);
-    assertChangesAllowForTargetLabel(commitsInPr, targetLabel, config);
-    assertCorrectBreakingChangeLabeling(commitsInPr, labels, config);
-  } catch (error) {
-    return error;
+    try {
+      assertPendingState(prData);
```
Comment (devversion) [inline ng-dev/pr/merge/pull-request.ts:93]: I feel like the pending state and correct breaking change stuff is unrelated to the labeö
config and should still run?

File: ng-dev/pr/merge/target-label.ts [hunk 2]
```diff
@@ -30,6 +31,12 @@ export function getTargetLabelFromPullRequest(
   config: Pick<MergeConfig, 'labels'>,
   labels: string[],
 ): TargetLabel {
+  if (config.labels === false) {
```
Comment (devversion) [inline ng-dev/pr/merge/target-label.ts:34]: Nit: I think it would be cleaner to create the labels config when this method is invoked instead. This method shouldn't really fetch the ng-dev config or deal with the defaults IMO.

Comment (devversion) [reply ng-dev/pr/merge/target-label.ts:34]: Also, the git client with the main branch name would exist there already

File: ng-dev/pr/merge/pull-request.ts (hunk 1 above)
Comment (josephperrott) [reply ng-dev/pr/merge/pull-request.ts:93]: Good call.

File: ng-dev/pr/merge/target-label.ts (hunk 2 above)
Comment (josephperrott) [reply ng-dev/pr/merge/target-label.ts:34]: Would it make sense instead to change the config to drop `labels` and add `noBranchTargeting?: boolean`

Then branch targeting can turned off entirely, so the options become either use the common branch targeting or just perform merges to wherever the PR targets in Github?

Comment (devversion) [reply ng-dev/pr/merge/target-label.ts:34]: great idea 👍

**Comment by josephperrott** [comment]
@devversion I kind of redid everything to move to having a property `noTargetLabeling` instead.

This is now built on top of #170

File: ng-dev/pr/merge/defaults/labels.ts [hunk 3]
```diff
@@ -32,15 +33,25 @@ import {assertActiveLtsBranch} from './lts-branch';
...
  */
-export async function getDefaultTargetLabelConfiguration(
-  api: GithubClient,
-  githubConfig: GitClientConfig,
-  releaseConfig: ReleaseConfig,
+export async function getTargetLabels(): Promise<TargetLabel[]>;
+export async function getTargetLabels(
+  _api?: GithubClient,
+  _config?: Partial<{github: GitClientConfig; release: ReleaseConfig}>,
+): Promise<TargetLabel[]>;
```
Comment (devversion) [inline ng-dev/pr/merge/defaults/labels.ts:40]: Are these overloads really needed here? Seems like the implementation signature already allows for these? We should remove them for readability.

File: ng-dev/pr/merge/defaults/labels.ts [hunk 4]
```diff
@@ -32,15 +33,25 @@ import {assertActiveLtsBranch} from './lts-branch';
...
+export async function getTargetLabels(
+  _api?: GithubClient,
+  _config?: Partial<{github: GitClientConfig; release: ReleaseConfig}>,
 ): Promise<TargetLabel[]> {
-  const nextBranchName = getNextBranchName(githubConfig);
+  const api = _api || GitClient.get().github;
+
+  const config = _config || (getConfig() as any);
```
Comment (devversion) [inline ng-dev/pr/merge/defaults/labels.ts:47]: We should use `??` instead of `||` for falling back (for `_api` as well). You could also inline the `GitClient.get().github` as default parameter IIUC. Maybe we should have a small local function here that does the same for the config (with the both assertions; so we do not need the weird `as any` cast here)

File: ng-dev/pr/merge/pull-request.ts [hunk 5]
```diff
@@ -73,22 +73,37 @@ export async function loadAndValidatePullRequest(
...
+
+  const githubTargetBranch = prData.baseRefName;
   let targetLabel: TargetLabel;
-  try {
-    targetLabel = getTargetLabelFromPullRequest(config, labels);
-  } catch (error) {
-    if (error instanceof InvalidTargetLabelError) {
-      return new PullRequestFailure(error.failureMessage);
+  let targetBranches: string[];
+  if (!config.noTargetLabeling) {
```
Comment (devversion) [inline ng-dev/pr/merge/pull-request.ts:82]: I'd probably swap the clauses here so that we do not double-negate.

File: ng-dev/pr/merge/defaults/labels.ts (hunk 4 above)
Comment (devversion) [reply ng-dev/pr/merge/defaults/labels.ts:47]: While reading below where `getTargetLabels` is called, I think we should just always require people to pass a `GithubClient` and the `config`. That is less confusing and more readable. Thoughts?

File: ng-dev/pr/merge/pull-request.ts (hunk 5 above)
Comment (devversion) [reply ng-dev/pr/merge/pull-request.ts:82]: Also, how about extracting this into a dedicated function? That would also allow us to use that for testing, compared to manually having to handle the `noTargetLabeling` case in tests (this would give us more coverage).

**Review by devversion — APPROVED** [review]
LGTM. A few minor suggestions

File: ng-dev/pr/merge/defaults/labels.ts (hunk 3 above)
Comment (josephperrott) [reply ng-dev/pr/merge/defaults/labels.ts:40]: The overloads are just in place so that we can provide the values in the `integration.spec.ts` tests.

File: ng-dev/pr/merge/defaults/labels.ts (hunk 4 above)
Comment (josephperrott) [reply ng-dev/pr/merge/defaults/labels.ts:47]: I initially had them as default parameters, but when it was a default parameter I ran into issues in the tests, but now they aren't happening so it must have been something else.

Comment (devversion) [reply ng-dev/pr/merge/defaults/labels.ts:47]: I see, yeah I think just making them required is probably the best IMO.

Comment (josephperrott) [reply ng-dev/pr/merge/defaults/labels.ts]: I am going to hold off on making them required as we are just moving the problem up a level in terms of the call stack by making it required here.  I think there is a larger refactoring that can address it.

Comment (devversion) [reply ng-dev/pr/merge/defaults/labels.ts]: I think it would be expected to pass the config around, compared to always retrieving it through the single-ton, but it's true that this is probably just up to personal preference. 

Yeah, let's keep it as is and we can see how it evolves over time.

File: ng-dev/pr/merge/config.ts [hunk 6]
```diff
@@ -85,49 +84,29 @@ export interface MergeConfig {
  * the dev-infra configuration is loaded as that could slow-down other commands.
```
Comment (devversion) [inline ng-dev/pr/merge/config.ts:84]: @josephperrott can we update that comment? it might be worth preserving the contents in some way, but saying that we do not support custom configuration, not sure. can be a follow-up too.

File: ng-dev/pr/merge/defaults/integration.spec.ts [hunk 7]
```diff
@@ -30,6 +31,7 @@ describe('default target labels', () => {
     api = new GithubClient();
     githubConfig = {owner: 'angular', name: 'dev-infra-test', mainBranchName: 'master'};
     releaseConfig = {
+      releaseNotes: {},
```
Comment (devversion) [inline ng-dev/pr/merge/defaults/integration.spec.ts:34]: Nit: this is no longer needed. `releaseNotes` is an option config.

```suggestion
```

File: ng-dev/pr/merge/pull-request.ts [hunk 8]
```diff
@@ -290,3 +273,32 @@ function assertPendingState(pr: RawPullRequest) {
       throw PullRequestFailure.isMerged();
   }
 }
+
+/** Get the branches the pull request will be merged into.  */
+async function getTargetBranches(
+  config: {merge: MergeConfig; github: GithubConfig},
+  labels: string[],
+  targetBranch: string,
```
Comment (devversion) [inline ng-dev/pr/merge/pull-request.ts:281]: We should keep the `githubTargetBranch` terminology to avoid ambiguity with actual target branches being determined. it's obviously reasonable from the context, but doesn't hurt staying more explicit.

```suggestion
  githubTargetBranch: string,
```

File: ng-dev/pr/merge/defaults/integration.spec.ts [hunk 9]
```diff
@@ -104,15 +106,17 @@ describe('default target labels', () => {
...
+    const labels = await computeTargetLabels();
+    spyOn(labelDefaults, 'getTargetLabels').and.resolveTo(labels);
     let label: TargetLabel;
     try {
-      label = getTargetLabelFromPullRequest({labels}, [name]);
+      label = await getTargetLabelFromPullRequest({}, [name]);
     } catch (error) {
+      console.log(error);
+      if (error.message === 'This repository does not use target labels') {
+        return [githubTargetBranch];
```
Comment (devversion) [inline ng-dev/pr/merge/defaults/integration.spec.ts:118]: This should be removed now. I think it's a dead path. We should fully rely on the actual logic from the merge script.

File: ng-dev/pr/merge/defaults/labels.ts (hunk 3 above)
Comment (devversion) [reply ng-dev/pr/merge/defaults/labels.ts:40]: I don't follow how these all overloads combined, are different from just having a single signature like:

```ts
export async function getTargetLabels(
  api = GitClient.get().github,
  config = getConfig() as Partial<{github: GithubConfig; release: ReleaseConfig}>,
): Promise<TargetLabel[]> {}

// both still work; regardless of overloads
getTargetLabels();
getTargetLabels(githubApi, config);
```

**Review by devversion — APPROVED** [review]
LGTM. A few minor comments, but looks great overall.

File: ng-dev/pr/merge/defaults/labels.ts (hunk 3 above)
Comment (josephperrott) [reply ng-dev/pr/merge/defaults/labels.ts:40]: I though I had removed the overloads when I was able to use the default parameters.  Fixed.

File: ng-dev/pr/merge/config.ts (hunk 6 above)
Comment (josephperrott) [reply ng-dev/pr/merge/config.ts:84]: I just removed it entirely as we no longer even need that type.
